kwasm-lemmas.md

```diff
@@ -308,68 +310,50 @@ We want to make the variant explicit, so we introduce the following helper, whic
-
-```k
-    rule #setRange(BM, EA, #getRange(BM, EA, WIDTH), WIDTH) => BM
-    rule #setRange(BM, EA, VAL, WIDTH) => #set(BM, EA, #wrap(1, VAL))
-      requires WIDTH ==Int 1
-      [simplification]
-
-    rule #getRange(BM, ADDR, WIDTH) modInt BYTE_SIZE => #get(BM, ADDR)
-      requires BYTE_SIZE ==Int 256
-       andBool notBool (WIDTH <=Int 0)
+    rule 0 <=Int #getRange(_, _, _) => true                                            [simplification]
+    rule 0 <=Int VAL <<Int SHIFT    => true requires 0 <=Int VAL andBool 0 <=Int SHIFT [simplification]
+    rule 0 <=Int VAL1 +Int VAL2     => true requires 0 <=Int VAL1 andBool 0 <=Int VAL2 [simplification]
+
+    rule #getRange(_, _, WIDTH)             <Int MAX => true requires 2 ^Int (8 *Int WIDTH) <=Int MAX                                       [simplification]
+    rule #getRange(_, _, WIDTH) <<Int SHIFT <Int MAX => true requires 2 ^Int ((8 *Int WIDTH) +Int SHIFT) <=Int MAX                          [simplification]
+    rule VAL1 +Int VAL2                     <Int MAX => true requires VAL1 <Int MAX andBool VAL2 <Int MAX andBool #distinctBits(VAL1, VAL2) [simplification]
+
+    rule #getRange(BM, ADDR, WIDTH) >>Int SHIFT => #getRange(BM, ADDR +Int 1, WIDTH -Int 1) >>Int (SHIFT -Int 8)
+      requires 0 <=Int ADDR
+       andBool 0  <Int WIDTH
+       andBool 8 <=Int SHIFT
        andBool #isByteMap(BM)
       [simplification]
 
-    rule #wrap(N, #getRange(BM, ADDR, WIDTH)) => #get(BM, ADDR)
-      requires N ==Int 1
-       andBool notBool (WIDTH <=Int 0)
+    rule #getRange(BM, ADDR, WIDTH) modInt MAX => #getRange(BM, ADDR, WIDTH -Int 1) modInt MAX
+      requires 0 <Int MAX andBool 0 <Int WIDTH
+       andBool 2 ^Int (8 *Int (WIDTH -Int 1)) modInt MAX ==Int 0
        andBool #isByteMap(BM)
       [simplification]
 
-    rule #wrap(N, #getRange(BM, ADDR, WIDTH)) => #getRange(BM, ADDR, N)
-      requires 0 <=Int N andBool N <=Int WIDTH
+    rule #wrap(MAX_WIDTH, #getRange(BM, ADDR, WIDTH)) => #getRange(BM, ADDR, minInt(MAX_WIDTH, WIDTH))
+      requires #isByteMap(BM)
       [simplification]
 
-    rule #getRange(BM, ADDR, WIDTH) => #get(BM, ADDR)
-      requires WIDTH ==Int 1
-      [simplification]
+    syntax Bool ::= #distinctBits ( Int , Int ) [function]
+ // ------------------------------------------------------
+    rule #distinctBits(#getRange(_, _, WIDTH), #getRange(_, _, _) <<Int SHIFT) => true requires WIDTH *Int 8 <=Int SHIFT
 ```
 
 `#getRange` over `#setRange`
 
 ```k
+    rule #setRange(BM, ADDR, #getRange(BM, ADDR, WIDTH), WIDTH) => BM [simplification]
+    rule #setRange(BM, ADDR, (#getRange(_, _, WIDTH1) #as VAL1) +Int (VAL2 <<Int SHIFT), WIDTH)
+      => #setRange(#setRange(BM, ADDR, VAL1, minInt(WIDTH1, WIDTH)), ADDR +Int WIDTH1, VAL2, WIDTH -Int WIDTH1)
+      requires 0 <=Int ADDR
+       andBool 0  <Int WIDTH
+       andBool WIDTH1 *Int 8 ==Int SHIFT
+      [simplification]
```

Comment:
Here we could technically do `ADDR +Int minInt(WIDTH1, WIDTH)`, and `WIDTH -Int minInt(WIDTH1, WIDTH)`, but it actually ends up being equivalent.

The `minInt` inside the inner `#setRange` is necessary though.

Reply:
Yep, good thinking.
Somehow it feels like it would be clearer if it was `ADDR +Int  SHIFT /Int 8` and `WIDTH -Int SHIFT /Int 8)`, but that's a bit larger, so I'm not sure if that's really better. But the outer `#setRange` really is not related to the `#getRange` value, only incidentally so because `WIDTH1 *Int 8 ==Int SHIFT`.